test/StreamJsonRpc.Tests/MarshalableProxyTests.cs

```diff
@@ -0,0 +1,621 @@
+                "someMethod",
+                new object?[] { marshalable, new JsonRpcTests.TypeThrowsWhenSerialized() },
+                new Type[] { typeof(IMarshalable), typeof(JsonRpcTests.TypeThrowsWhenSerialized) },
+                this.TimeoutToken));
+            Assert.True(ex is JsonSerializationException || ex is MessagePackSerializationException);
+            Assert.True(IsExceptionOrInnerOfType<Exception>(ex, exactTypeMatch: true));
+
+            return new WeakReference(marshalable);
+        }
+    }
+
+    [Fact]
+    public async Task NoLeakWhenServerThrows()
+    {
+        WeakReference weakRef = await Helper();
+        await this.AssertWeakReferenceGetsCollectedAsync(weakRef);
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        async Task<WeakReference> Helper()
+        {
+            var marshalable = new Data();
+            await Assert.ThrowsAsync<RemoteMethodNotFoundException>(() => this.clientRpc.InvokeWithCancellationAsync(
+                "someMethod",
+                new object?[] { marshalable },
+                new Type[] { typeof(IMarshalable) },
+                this.TimeoutToken));
+            return new WeakReference(marshalable);
+        }
+    }
+
+    [Fact]
+    public async Task IMarshalableInNotificationArgumentIsRejected()
+    {
+        var ex = await Assert.ThrowsAnyAsync<Exception>(() => this.clientRpc.NotifyAsync("someMethod", new object?[] { new Data() }, new Type[] { typeof(IMarshalable) }));
+        Assert.True(IsExceptionOrInnerOfType<NotSupportedException>(ex));
+    }
+
+    [Fact]
+    public async Task MarshalableInterfaceMustBeDisposable()
+    {
+        var ex = await Assert.ThrowsAnyAsync<Exception>(() => this.client.AcceptNonDisposableMarshalableAsync(new NonDisposableMarshalable()));
+        Assert.True(IsExceptionOrInnerOfType<NotSupportedException>(ex));
+    }
+
+    [Fact]
+    public async Task MarshalableInterfaceCannotHaveProperties()
+    {
+        var ex = await Assert.ThrowsAnyAsync<Exception>(() => this.client.AcceptMarshalableWithPropertiesAsync(new MarshalableWithProperties()));
+        Assert.True(IsExceptionOrInnerOfType<NotSupportedException>(ex));
+    }
+
+    [Fact]
+    public async Task MarshalableInterfaceCannotHaveEvents()
+    {
+        var ex = await Assert.ThrowsAnyAsync<Exception>(() => this.client.AcceptMarshalableWithEventsAsync(new MarshalableWithEvents()));
+        Assert.True(IsExceptionOrInnerOfType<NotSupportedException>(ex));
+    }
+
+    [Fact]
+    public async Task InterfacesMustBeMarkedAsRpcMarshalable()
+    {
+        await Assert.ThrowsAnyAsync<Exception>(() => this.client.AcceptNonMarshalableAsync(new NonDataContractMarshalable()));
+    }
+
+    [Fact]
+    public async Task RpcMarshalableAttributeDoesntAffectDerivedInterfaces()
+    {
+        await Assert.ThrowsAnyAsync<Exception>(() => this.client.AcceptNonMarshalableDerivedFromMarshalablesAsync(new NonDataContractMarshalable()));
```

Comment:
Why does this throw? 
What if the interface included attributes to instruct the formatter as to how to format it as data, the way interfaces can typically do? I would expect that to work, so it would be interesting to see a test to show it.

Reply:
You are saying that an interface could be marked as `[DataContract]`? If that is supported, I will add tests for that.

But what do we do in case an interface is both `[DataContract]` and `RpcMarshalable`? Which one wins?

About why this throws: `AcceptNonMarshalableDerivedFromMarshalablesAsync` takes a parameter of type `INonMarshalableDerivedFromMarshalable` where
```cs
    public interface INonMarshalableDerivedFromMarshalable : IMarshalable
    {
    }

    [RpcMarshalable]
    public interface IGenericMarshalable<T> : IMarshalable
    {
        Task<T> DoSomethingWithParameterAsync(T paremeter);
    }
```

We don't allow to use a derived interface type as if it were also `RpcMarshalable`.

This is reinforced by `RpcMarshalableAttribute` being marked as `Inherited = false`.

Reply:
We clarified in person that an interface could be both `[RpcMarshalable]` and otherwise serializable (e.g., through `[MessagePackFormatter]` or `[JsonConverter]`) and that the desired behavior would be either throwing an exception or marshaling.

I changed the implementation to marshal these interfaces (I fear that the check necessary to throw an exception would lead to negative performance impact).

For MessagePack, the solution is simply to move `RpcMarshalableImplicitResolver` from the bottom to the top of the resolver list.

For Newtonsoft.Json, the solution is to leverage the contract resolver.